btree/immutable/cacher.go

```diff
@@ -0,0 +1,114 @@
+package btree
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Workiva/go-datastructures/futures"
+)
+
+// cacher provides a convenient interface for retrieving,
+// storing, and caching nodes.  This ensures that we don't have to constantly
+// run to the persister to fetch nodes we are using over and over again.
+// TODO: this should probably evict items from the cache if the cache gets
+// too full.
+type cacher struct {
+	lock      sync.Mutex
+	cache     map[string]*futures.Future
+	persister Persister
+}
+
+// clear deletes all items from the cache.
+func (c *cacher) clear() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.cache = make(map[string]*futures.Future, 10)
+}
+
+// deleteFromCache will remove the provided ID from the cache.  This
+// is a threadsafe operation.
+func (c *cacher) deleteFromCache(id ID) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	delete(c.cache, string(id))
+}
+
+func (c *cacher) loadNode(t *Tr, key ID) (*Node, error) {
+	items, err := c.persister.Load(key)
+	if err != nil {
+		return nil, err
+	}
+
+	n, err := nodeFromBytes(t, items[0].Payload)
+	if err != nil {
+		return nil, err
+	}
+
+	return n, nil
+}
+
+// getNode will return a Node matching the provided id.  An error is returned
+// if the cacher could not go to the persister or the node could not be found.
+// All found nodes are cached so subsequent calls should be faster than
+// the initial.  This blocks until the node is loaded, but is also threadsafe.
+func (c *cacher) getNode(t *Tr, key ID, cache bool) (*Node, error) {
```

Comment:
Personally, I'd rename the `cache` variable to `useCache` to more clearly express that it's a boolean and not an object.